### PR TITLE
Feature 15348 - Quick Access Saved Filters

### PR DESCRIPTION
**Pull request by Julio-Oliveira-Encora**
### Feature: #15348

A dropdown was added to list and apply the saved filters to the current view.

### CODE REVIEW

File: netbox/templates/inc/table_controls_htmx.html [hunk 1]
```diff
@@ -4,22 +4,34 @@
 <div class="row mb-3">
   <div class="col-auto d-print-none">
     <div class="input-group input-group-flat me-2 quicksearch" hx-disinherit="hx-select hx-swap">
-      <input type="search" results="5" name="q" id="quicksearch" class="form-control px-2 py-1" placeholder="Quick search"
-        hx-get="{{ request.full_path }}" hx-target="#object_list" hx-trigger="keyup changed delay:500ms, search" />
+      <input type="search" results="5" name="q" id="quicksearch" class="form-control px-2 py-1"
```
Comment (jeremystretch) [inline netbox/templates/inc/table_controls_htmx.html:7]: We need to figure out how to make the new dropdown visually consistent with the quick search box.

screenshot

File: netbox/project-static/src/savedFiltersSelect.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Handle saved filter change event.
+ *
+ * @param event "change" event for the saved filter select
+ */
+function handleSavedFilterChange(event: Event): void {
+  const savedFilter = event.currentTarget as HTMLSelectElement;
+  let baseUrl = savedFilter.baseURI.split("?")[0];
+  let preFilter = "?";
```
Comment (jeremystretch) [inline netbox/project-static/src/savedFiltersSelect.ts:11]:
```suggestion
  const preFilter = '?';
```

File: netbox/project-static/src/savedFiltersSelect.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import {isTruthy} from "./util";
```
Comment (jeremystretch) [inline netbox/project-static/src/savedFiltersSelect.ts:1]: This and some other lines within the file trigger complaints from `prettier`.

```suggestion
import { isTruthy } from './util';
```

File: netbox/project-static/src/savedFiltersSelect.ts
Comment (jeremystretch) [inline netbox/project-static/src/savedFiltersSelect.ts:1]: We should probably move this under `forms/`.

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Although functional, displaying the active saved filters in both the original tag form and in the selector widget may be confusing to the user. IMO we should display applied filters only in the selector widget, and show only individual filter values as removable colored tags.

screenshot

**Comment by arthanson** [comment]
@Julio-Oliveira-Encora here are the changes after discussion on this:

* Remove the blue chits for the saved filters only, so normal filters will still show there (The functionality for saving your filters and such should remain).  So the selected saved filters will only show in the save filters dropdown.
* We don't need to change the saved filters to use a placeholder, having the label there is fine.
* Make the vertical height of the dropdown match the Quick Search so at start they aren't different sizes.  You can adjust the text height and padding of the Saved filters, you may also need to tweak the quick search if needed.  Note: If you select many saved filters it will wrap and grow the select box, this is fine - we just care about the initial page look when you have no filters (or one filter) selected.

**Comment by Julio-Oliveira-Encora** [comment]
@arthanson I changed the CSS for quick search, because I did get to change the CSS for Dropdown. Please, check them.

**Review by arthanson — CHANGES_REQUESTED** [review]
@Julio-Oliveira-Encora I added a saved filter and then selected it, however it didn't actually filter the table - see screenshots.  The first is the filter applied manually, the second is selecting it from the saved filter list, they number of results should match but don't.
Monosnap IP Addresses | NetBox 2024-05-28 09-19-39
Monosnap IP Addresses | NetBox 2024-05-28 09-20-03

**Comment by arthanson** [comment]
@Julio-Oliveira-Encora I just re-tested this, it is happening when I don't have HTMX navigation enabled

**Comment by Julio-Oliveira-Encora** [comment]
I tested with HTMX Navigation disabled:

Preferences


For IP Address I created a "Saved Filter" for IPV4 and Active. It worked:

ip_saved_filter
ip_filters


For sites I tested too:

sites
Screenshot from 2024-06-03 12-17-17

**Comment by arthanson** [comment]
@Julio-Oliveira-Encora almost there - can you remove the saved-filters from the filters tab. With both the dropdown and it being in filters tab you can get the UI showing both which isn't great:
Monosnap Devices | NetBox 2024-06-07 08-47-39

File: netbox/templates/inc/table_controls_htmx.html [hunk 4]
```diff
@@ -1,25 +1,37 @@
...
       <span class="input-group-text py-1">
         <a href="#" id="quicksearch_clear" class="invisible text-secondary"><i class="mdi mdi-close-circle"></i></a>
       </span>
-    {% block extra_table_controls %}{% endblock %}
+      {% block extra_table_controls %}{% endblock %}
     </div>
   </div>
+
+  <div class="col-auto d-print-none">
+    <label class="my-2" for="id_filter_id">Saved Filter</label>
```
Comment (arthanson) [inline netbox/templates/inc/table_controls_htmx.html:18]: Also put {% trans around this

File: netbox/project-static/src/appliedFilter.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ */
+export function initAppliedFilter(): void {
+  const appliedFilter = document.getElementById('appliedfilters');
+  if (isTruthy(appliedFilter)) {
+    const divResults = document.getElementById('results');
+    if (isTruthy(divResults)) {
+      const savedFilterSelect = divResults.getElementsByTagName('select')[0];
+      if (savedFilterSelect.selectedOptions.length > 0) {
+        appliedFilter.hidden = true;
```
Comment (jeremystretch) [inline netbox/project-static/src/appliedFilter.ts:14]: We should not rely on Javascript to hide this element; rather it should simply not be rendered in the template.

@arthanson can you help Julio with this please?

**Review by jeremystretch — CHANGES_REQUESTED** [review]

**Comment by arthanson** [comment]
@Julio-Oliveira-Encora if you need any help with hiding the column in the table, please let me know.

**Comment by Julio-Oliveira-Encora** [comment]
@arthanson  and @jeremystretch  
Could the function `applied_filters` be modified to check if the form only contains the  `filter_id` and return an empty `applied_filters` to avoid using Javascript?

**Review by jeremystretch — APPROVED** [review]
I've omitted saved filters from the "applied filters" list and did some cosmetic cleanup of the selector widget.

**Comment by Julio-Oliveira-Encora** [comment]
@jeremystretch Should I make some additional changes to it?

**Review by arthanson — APPROVED** [review]

**Pull request merged**